File: examples/jvm/src/main/java/com/example/chatbot/Bot.java
package com.example.chatbot;

import com.fasterxml.jackson.databind.JsonNode;
import com.x.chatxdk.Types.SigningKeyEntry;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * The receive -&gt; decrypt -&gt; reply -&gt; encrypt -&gt; send loop. Conversation state
 * is kept in memory, one entry per conversation.
 */
public final class Bot {

    private final ChatCore core;
    private final XChatClient api;
    private final String botUserId;
    private final Map<String, ConversationState> state = new HashMap<>();

    public Bot(ChatCore core, XChatClient api, String botUserId) {
        this.core = core;
        this.api = api;
        this.botUserId = botUserId;
        // Session identity, once: every encrypt below signs as the bot without
        // passing a sender id per call.
        core.setIdentity(botUserId);
    }

    /** Turn an incoming message into a reply (simple echo by default). */
    public static String generateReply(String text) {
        String t = text.trim();
        return (t.equals("ping") || t.equals("!ping")) ? "pong" : "You said: " + t;
    }

    private ConversationState stateFor(String conversationId) {
        return state.computeIfAbsent(conversationId, k -> new ConversationState());
    }

    private List<SigningKeyEntry> signingKeysFor(List<JsonNode> events) throws Exception {
        Set<String> senders = new LinkedHashSet<>();
        for (JsonNode e : events) {
            String sid = e.path("sender_id").asText("");
            if (!sid.isEmpty() && !sid.equals(botUserId)) senders.add(sid);
        }
        List<SigningKeyEntry> keys = new ArrayList<>();
        for (String senderId : senders) {
            try {
                for (JsonNode pk : api.getPublicKeys(senderId)) {
                    SigningKeyEntry entry = new SigningKeyEntry();
                    entry.userId = senderId;
                    entry.publicKeyVersion = pk.path("public_key_version").asText("");
                    entry.publicKey = pk.path("signing_public_key").asText("");
                    entry.identityPublicKey = pk.path("public_key").asText("");
                    entry.identityPublicKeySignature = pk.path("identity_public_key_signature").asText("");
                    keys.add(entry);
                }
            } catch (Exception ex) {
                System.err.println("public_keys_fetch_failed sender=" + senderId);
            }
        }
        return keys;
    }

    private static List<JsonNode> dataArray(JsonNode page) {
        List<JsonNode> out = new ArrayList<>();
        JsonNode data = page.path("data");
        if (data.isArray()) data.forEach(out::add);
        return out;
    }

    /** Pagination token from a GET events page, or null when absent. */
    private static String nextToken(JsonNode page) {
        JsonNode nt = page.path("meta").path("next_token");
        return nt.isTextual() ? nt.asText() : null;
    }

    /** Initial load: batch-decrypt the backlog (decryptEvents path). */
    public void loadBacklog(String conversationId) throws Exception {
        ConversationState st = stateFor(conversationId);
        JsonNode page = api.getEvents(conversationId, 100, null);
        List<JsonNode> raw = dataArray(page);
        List<String> eventsB64 = new ArrayList<>();
        for (JsonNode e : raw) {
            String ev = e.path("encoded_event").asText("");
            if (!ev.isEmpty()) eventsB64.add(ev);
        }
        var result = core.decryptBatch(eventsB64, signingKeysFor(raw));
        st.conversationKeys.putAll(result.conversationKeys.keys);
        st.latestKeyVersion = result.conversationKeys.latestVersion;
        st.paginationToken = nextToken(page);
        System.out.printf("backlog_loaded conv=%s messages=%d keys=%d%n",
                conversationId, result.messages.size(), result.conversationKeys.keys.size());
    }

    /** Poll for new events; reply to each new message (decryptEvent path). */
    public void pollOnce(String conversationId) throws Exception {
        ConversationState st = stateFor(conversationId);
        JsonNode page = api.getEvents(conversationId, 50, st.paginationToken);
        List<JsonNode> raw = dataArray(page);
        List<SigningKeyEntry> signingKeys = signingKeysFor(raw);

        for (JsonNode item : raw) {
            String eventB64 = item.path("encoded_event").asText("");
            if (eventB64.isEmpty()) continue;
            JsonNode event = core.decryptOne(eventB64, st.conversationKeys, signingKeys);

            if ("KeyChange".equals(event.path("type").asText())) {
                String keyVersion = event.path("key_version").asText();
                for (JsonNode pk : event.path("participant_keys")) {
                    String enc = pk.path("encrypted_key").asText("");
                    if (enc.isEmpty()) continue;
                    try {
                        st.conversationKeys.put(keyVersion, core.decryptConversationKey(enc));
                        st.latestKeyVersion = keyVersion;
                        break;
                    } catch (RuntimeException ignored) {
                        // not for us
                    }
                }
                continue;
            }
            maybeReply(conversationId, event, eventB64);
        }
        // Advance the pagination token so the next poll fetches only new events.
        String next = nextToken(page);
        if (next != null && !next.isEmpty()) st.paginationToken = next;
    }

    private void maybeReply(String conversationId, JsonNode event, String eventB64) throws Exception {
        ConversationState st = stateFor(conversationId);
        String eventId = event.path("id").asText("");
        String senderId = event.path("sender_id").asText("");
        if (eventId.isEmpty() || !st.seenEventIds.add(eventId)) return;
        if (senderId.equals(botUserId)) return;

        String text = ChatCore.messageText(event);
        if (text == null || text.isEmpty()) return;

        String keyVersion = event.path("key_version").asText(st.latestKeyVersion);
        byte[] convKey = keyVersion == null ? null : st.conversationKeys.get(keyVersion);
        if (convKey == null || keyVersion == null) {
            System.err.println("no_conversation_key conv=" + conversationId);
            return;
        }

        // The message signature covers the conversation_id, so sign with the
        // canonical id carried inside the event (the X API uses a different
        // separator in its URL paths than the form embedded in events).
        String replyConvId = event.path("conversation_id").asText(conversationId);
        String reply = generateReply(text);
        // Reply by raw event: the SDK derives the threaded-reply preview from
        // the incoming signed event and embeds it for recipient validation.
        ChatCore.SendBody body =
                core.encryptReply(replyConvId, reply, convKey, keyVersion, eventB64, null, null, null, null);
        api.sendMessage(replyConvId, body);
        System.out.printf("reply_sent conv=%s len=%d%n", replyConvId, reply.length());
    }

    /** Load the backlog then poll forever. */
    public void run(String conversationId, long pollIntervalMs) throws Exception {
        loadBacklog(conversationId);
        System.out.printf("bot_running conv=%s polling every %dms%n", conversationId, pollIntervalMs);
        while (true) {
            try {
                pollOnce(conversationId);
            } catch (Exception e) {
                System.err.println("poll_error conv=" + conversationId + " " + e.getMessage());
            }
            Thread.sleep(pollIntervalMs);
        }
    }

    private static final class ConversationState {
        final Map<String, byte[]> conversationKeys = new HashMap<>();
        String latestKeyVersion;
        final Set<String> seenEventIds = new HashSet<>();
        String paginationToken;
    }
}

File: examples/dotnet/ChatBot/Bot.cs
using System.Text.Json;
using ChatXdk;

namespace ChatBot;

/// <summary>
/// The receive -> decrypt -> reply -> encrypt -> send loop. Conversation state
/// is kept in memory, one entry per conversation.
/// </summary>
public sealed class Bot
{
    private readonly ChatCore _core;
    private readonly XChatClient _api;
    private readonly string _botUserId;
    private readonly Dictionary<string, ConversationState> _state = new();

    public Bot(ChatCore core, XChatClient api, string botUserId)
    {
        _core = core;
        _api = api;
        _botUserId = botUserId;
        // Session identity, once: every encrypt below signs as the bot without
        // passing a sender id per call.
        _core.SetIdentity(botUserId);
    }

    /// <summary>Turn an incoming message into a reply (simple echo by default).</summary>
    public static string GenerateReply(string text) =>
        text.Trim() is "ping" or "!ping" ? "pong" : $"You said: {text.Trim()}";

    private ConversationState State(string conversationId)
    {
        if (!_state.TryGetValue(conversationId, out var st))
        {
            st = new ConversationState();
            _state[conversationId] = st;
        }
        return st;
    }

    /// <summary>Pagination token from a GET events page, or null when absent.</summary>
    private static string? NextToken(JsonElement page) =>
        page.TryGetProperty("meta", out var meta)
        && meta.ValueKind == JsonValueKind.Object
        && meta.TryGetProperty("next_token", out var nt)
        && nt.ValueKind == JsonValueKind.String
            ? nt.GetString()
            : null;

    /// <summary>
    /// KeyChange events from a GET events page. They arrive in
    /// meta.conversation_key_events, separate from data, and carry the
    /// conversation keys — they must go into the same DecryptEvents batch as
    /// the data events.
    /// </summary>
    private static List<string> KeyEvents(JsonElement page) =>
        page.TryGetProperty("meta", out var meta)
        && meta.ValueKind == JsonValueKind.Object
        && meta.TryGetProperty("conversation_key_events", out var arr)
        && arr.ValueKind == JsonValueKind.Array
            ? arr.EnumerateArray()
                .Where(e => e.ValueKind == JsonValueKind.String)
                .Select(e => e.GetString()!)
                .ToList()
            : new List<string>();

    private async Task<List<SigningKeyEntry>> SigningKeysForAsync(IEnumerable<JsonElement> events)
    {
        var senders = events
            .Select(e => e.TryGetProperty("sender_id", out var s) ? s.GetString() : null)
            .Where(id => !string.IsNullOrEmpty(id) && id != _botUserId)
            .Distinct()
            .ToList();

        var keys = new List<SigningKeyEntry>();
        foreach (var senderId in senders)
        {
            try
            {
                foreach (var pk in await _api.GetPublicKeysAsync(senderId!))
                {
                    keys.Add(new SigningKeyEntry
                    {
                        UserId = senderId!,
                        PublicKeyVersion = pk.TryGetProperty("public_key_version", out var v) ? v.GetString() ?? "" : "",
                        PublicKey = pk.TryGetProperty("signing_public_key", out var sp) ? sp.GetString() ?? "" : "",
                        IdentityPublicKey = pk.TryGetProperty("public_key", out var p) ? p.GetString() ?? "" : "",
                        IdentityPublicKeySignature = pk.TryGetProperty("identity_public_key_signature", out var sig) ? sig.GetString() ?? "" : "",
                    });
                }
            }
            catch
            {
                Console.Error.WriteLine($"public_keys_fetch_failed sender={senderId}");
            }
        }
        return keys;
    }

    /// <summary>Initial load: batch-decrypt the backlog (DecryptEvents path).</summary>
    public async Task LoadBacklogAsync(string conversationId)
    {
        var st = State(conversationId);
        var page = await _api.GetEventsAsync(conversationId, 100);
        var raw = page.TryGetProperty("data", out var d) && d.ValueKind == JsonValueKind.Array
            ? d.EnumerateArray().ToList()
            : new List<JsonElement>();
        var eventsB64 = KeyEvents(page);
        eventsB64.AddRange(raw
            .Select(e => e.TryGetProperty("encoded_event", out var ev) ? ev.GetString() : null)
            .Where(s => !string.IsNullOrEmpty(s))
            .Select(s => s!));

        var result = _core.DecryptBatch(eventsB64, await SigningKeysForAsync(raw));
        foreach (var (version, key) in result.ConversationKeys.Keys)
            st.ConversationKeys[version] = key;
        st.LatestKeyVersion = result.ConversationKeys.LatestVersion;
        st.PaginationToken = NextToken(page);
        Console.WriteLine($"backlog_loaded conv={conversationId} messages={result.Messages.Count} keys={result.ConversationKeys.Keys.Count}");
    }

    /// <summary>Poll for new events; reply to each new message (DecryptEvent path).</summary>
    public async Task PollOnceAsync(string conversationId)
    {
        var st = State(conversationId);
        var page = await _api.GetEventsAsync(conversationId, 50, st.PaginationToken);
        var raw = page.TryGetProperty("data", out var d) && d.ValueKind == JsonValueKind.Array
            ? d.EnumerateArray().ToList()
            : new List<JsonElement>();
        var signingKeys = await SigningKeysForAsync(raw);

        // Key changes for this page arrive in meta, not data; adopt their
        // keys before decrypting the messages that need them.
        var pageKeyEvents = KeyEvents(page);
        if (pageKeyEvents.Count > 0)
        {
            var rotated = _core.DecryptBatch(pageKeyEvents, signingKeys);
            foreach (var (version, key) in rotated.ConversationKeys.Keys)
                st.ConversationKeys[version] = key;
            if (rotated.ConversationKeys.LatestVersion is { } lv)
                st.LatestKeyVersion = lv;
        }

        foreach (var item in raw)
        {
            if (!item.TryGetProperty("encoded_event", out var ev) || ev.GetString() is not { } eventB64)
                continue;
            var decrypted = _core.DecryptOne(eventB64, st.ConversationKeys, signingKeys);

            if (decrypted.TryGetProperty("type", out var t) && t.GetString() == "KeyChange")
            {
                var keyVersion = decrypted.GetProperty("key_version").GetString()!;
                foreach (var pk in decrypted.GetProperty("participant_keys").EnumerateArray())
                {
                    var enc = pk.GetProperty("encrypted_key").GetString();
                    if (string.IsNullOrEmpty(enc)) continue;
                    try
                    {
                        st.ConversationKeys[keyVersion] = _core.DecryptConversationKey(enc);
                        st.LatestKeyVersion = keyVersion;
                        break;
                    }
                    catch { /* not for us */ }
                }
                continue;
            }
            await MaybeReplyAsync(conversationId, decrypted, eventB64);
        }
        // Advance the pagination token so the next poll fetches only new events.
        if (NextToken(page) is { Length: > 0 } next)
            st.PaginationToken = next;
    }

    private async Task MaybeReplyAsync(string conversationId, JsonElement evt, string eventB64)
    {
        var st = State(conversationId);
        var eventId = evt.TryGetProperty("id", out var i) ? i.GetString() ?? "" : "";
        var senderId = evt.TryGetProperty("sender_id", out var s) ? s.GetString() ?? "" : "";
        if (string.IsNullOrEmpty(eventId) || !st.SeenEventIds.Add(eventId)) return;
        if (senderId == _botUserId) return;

        var text = EventHelpers.MessageText(evt);
        if (string.IsNullOrEmpty(text)) return;

        var keyVersion = evt.TryGetProperty("key_version", out var kv) ? kv.GetString() : st.LatestKeyVersion;
        if (keyVersion is null || !st.ConversationKeys.TryGetValue(keyVersion, out var convKey))
        {
            Console.Error.WriteLine($"no_conversation_key conv={conversationId}");
            return;
        }

        // The message signature covers the conversation_id, so sign with the
        // canonical id carried inside the event (the X API uses a different
        // separator in its URL paths than the form embedded in events).
        var replyConvId = evt.TryGetProperty("conversation_id", out var cid)
            ? cid.GetString() ?? conversationId
            : conversationId;
        var reply = GenerateReply(text);
        // Reply by raw event: the SDK derives the threaded-reply preview from
        // the incoming signed event and embeds it for recipient validation.
        var body = _core.EncryptReply(replyConvId, reply, convKey, keyVersion, replyToEvent: eventB64);
        await _api.SendMessageAsync(replyConvId, body);
        Console.WriteLine($"reply_sent conv={replyConvId} len={reply.Length}");
    }

    /// <summary>Load the backlog then poll forever.</summary>
    public async Task RunAsync(string conversationId, TimeSpan pollInterval)
    {
        await LoadBacklogAsync(conversationId);
        Console.WriteLine($"bot_running conv={conversationId} polling every {pollInterval}");
        while (true)
        {
            try { await PollOnceAsync(conversationId); }
            catch (Exception e) { Console.Error.WriteLine($"poll_error conv={conversationId} {e.Message}"); }
            await Task.Delay(pollInterval);
        }
    }

    private sealed class ConversationState
    {
        public Dictionary<string, byte[]> ConversationKeys { get; } = new();
        public string? LatestKeyVersion { get; set; }
        public HashSet<string> SeenEventIds { get; } = new();
        public string? PaginationToken { get; set; }
    }
}
